### DISCUSSION TITLE
Useful debugging tool for workflows which have been run remotely?

### DISCUSSION THREAD

**Discussion by sminot** [nextflow-io/nextflow#2436] [Ideas]
When running Nextflow locally, it's really easy to debug a failure at a particular step. You can just:
- `cd` to the work directory
- Check the logs to see what the source of the error might be
- Identify the source of all input files by checking the source of the symlink (`ls -lah`)
- Navigate to the work directory of any upstream files
- Repeat as necessary

Doing the same thing for workflows run remotely (e.g., on AWS) is a lot harder. I'm finding myself doing a lot of:
- Paging through the log or trace file to figure out the work directory of any failed steps
- `aws s3 cp $WORK_DIR/.command.log - | less` to look at the logs a step
- `aws s3 cp $WORK_DIR/.command.run - | grep s3 | grep download` to find the inputs to that step

I don't know what the best implementation is, but some way to make it easier to perform these debugging activities would be really useful.

cc: @pditommaso

**Comment by pditommaso** [comment]
That's a good point. My first thought is that we should improve the Tower task detail including the list of input and output files or add the ability to browse the task work dir.

**Comment by sminot** [comment]
There's definitely a lot of utility in building out features in Tower, but another need to consider is the ability to rapidly parse input/output files with command line utilities. For example, if an error is caused by a file not being gzip-compressed appropriately, that is very easy to diagnose on the command line (for an expert-level user), but much less easy in the browser. It might be worth considering whether there is a command-line utility that could streamline this type of debugging.

**Reply by pditommaso** [reply]
Basically do you need to list the tasks input files?

**Reply by sminot** [reply]
The items I'm thinking about are:
- Quickly identifying any tasks which failed
- Viewing stdout/err
- Figuring out the full work directory path from the short task hash
- Listing input files for a task

**Reply by pditommaso** [reply]
Thinking more, I think making a Bash wrapper that launches a FUSE driver e.g. [Gooyfs](https://github.com/kahing/goofys) and jump into the bucket work dir could do be a nice workaround
